carma_guidance_plugins/include/carma_guidance_plugins/control_plugin.hpp

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (C) 2022 LEIDOS.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+#pragma once
+
+#include <rclcpp/rclcpp.hpp>
+#include <carma_planning_msgs/msg/trajectory_plan.hpp>
+#include <geometry_msgs/msg/pose_stamped.hpp>
+#include <geometry_msgs/msg/twist_stamped.hpp>
+#include <autoware_msgs/msg/control_command_stamped.hpp>
+
+#include "carma_guidance_plugins/plugin_base_node.hpp"
+
+namespace carma_guidance_plugins
+{
+
+  /**
+   * \brief ControlPlugin base class which can be extended by user provided plugins which wish to implement the Control Plugin ROS API. 
+   * 
+   * A control plugin is responsible for generating high frequency vehicle speed and steering commands to execute the currently planned trajectory. 
+   * This plugin provides default subscribers to track the pose, velocity, and current trajectory in the system. 
+   * Extending classes must implement the generate_command() method to use that data and or additional data to plan commands at a 30Hz frequency. 
+   * 
+   */
+  class ControlPlugin : public PluginBaseNode
+  {
+
+  private:
+    // Subscriptions
+    carma_ros2_utils::SubPtr<geometry_msgs::msg::PoseStamped> current_pose_sub_;
+    carma_ros2_utils::SubPtr<geometry_msgs::msg::TwistStamped> current_velocity_sub_;
+    carma_ros2_utils::SubPtr<carma_planning_msgs::msg::TrajectoryPlan> trajectory_plan_sub_;
+
+    // Publishers
+    carma_ros2_utils::PubPtr<autoware_msgs::msg::ControlCommandStamped> vehicle_cmd_pub_;
+
+    // Timers
+    rclcpp::TimerBase::SharedPtr command_timer_;
+
+    // These callbacks do direct assignment into their respective member variables
+    void current_pose_callback(geometry_msgs::msg::PoseStamped::UniquePtr msg);
+    void current_twist_callback(geometry_msgs::msg::TwistStamped::UniquePtr msg);
+    void current_trajectory_callback(carma_planning_msgs::msg::TrajectoryPlan::UniquePtr msg);
+
+
+  protected:
+
+    //! The most recent pose message received by this node
+    boost::optional<geometry_msgs::msg::PoseStamped> current_pose_;
```

Comment:
Is it worth documenting somewhere in here that the subscriptions/etc. are all setup to support only single-threaded plugins? We don't have any multi-threaded plugins but, in the event we needed to create one, I don't think these structures here are synchronized, so we'd need to start from a different base class.

Reply:
Yeah, I'll add some comments. I think it would be a good idea to provide multi-threading support but I don't think its an immediate need.